### PR TITLE
return 1 entry for joint tenants

### PR DESCRIPTION
**Pull request by elena-vi**

### CODE REVIEW

File: LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs [hunk 1]
```diff
@@ -33,10 +36,34 @@ public async Task<SearchTenancyResponse> ExecuteAsync(SearchTenancyRequest reque
             if (response == null)
                 return new SearchTenancyResponse();
 
+            var allTenancyListItems = response.Results;
+
+            var uniqueTenancyListItems = allTenancyListItems.GroupBy(t => t.TenancyRef)
+                .Where(g => g.Count() == 1)
+                .SelectMany(g => g).ToList();;
+
+            var groupsOfDuplicatesTenancyListItems =
```
Comment (maysakanoni) [inline LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs]:
```suggestion
            var groupsOfDuplicateTenancyListItems =
```

File: LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs [hunk 2]
```diff
@@ -33,10 +36,34 @@ public async Task<SearchTenancyResponse> ExecuteAsync(SearchTenancyRequest reque
...
+                allTenancyListItems.GroupBy(t => t.TenancyRef)
+                .Where(g => g.Count() > 1);
+
+            foreach (var grouping in groupsOfDuplicatesTenancyListItems)
```
Comment (maysakanoni) [inline LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs]:
```suggestion
            foreach (var grouping in groupsOfDuplicateTenancyListItems)
```

File: LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs [hunk 3]
```diff
@@ -33,10 +36,34 @@ public async Task<SearchTenancyResponse> ExecuteAsync(SearchTenancyRequest reque
             if (response == null)
                 return new SearchTenancyResponse();
 
+            var allTenancyListItems = response.Results;
+
+            var uniqueTenancyListItems = allTenancyListItems.GroupBy(t => t.TenancyRef)
+                .Where(g => g.Count() == 1)
+                .SelectMany(g => g).ToList();;
```
Comment (thip) [inline LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs]: don't think you need the extra `;` here 😉

File: LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs [hunk 4]
```diff
@@ -33,10 +36,34 @@ public async Task<SearchTenancyResponse> ExecuteAsync(SearchTenancyRequest reque
...
+
+            foreach (var grouping in groupsOfDuplicatesTenancyListItems)
+            {
+                var jointTenancies = grouping.ToList();
+                var jointTenancy = jointTenancies[0];
+                jointTenancies.ForEach(delegate(TenancyListItem dup)
+                {
+                    if (jointTenancy.PrimaryContactName != dup.PrimaryContactName)
+                    {
+                        jointTenancy.PrimaryContactName += " & " + dup.PrimaryContactName;
```
Comment (thip) [inline LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs]: might want to think about using string interpolation here:

`jointTenancy.PrimaryContactName = $"{jointTenancy.PrimaryContactName} & {dup.PrimaryContactName}"`

(might also be a better way!)

**Comment by cob16** [comment]
This solves the problem at the application level. :+1:  It might be worth considering instead adding `SELECT DISTINCT` to the SQL query. This would make the DB not return duplicates instead

**Comment by elena-vi** [comment]
Can't do `SELECT DISTINCT` because the officers want to see joint tenancies

File: LBHTenancyAPITest/Test/UseCases/V2/Search/SearchTenancyUseCaseTests.cs [hunk 5]
```diff
@@ -103,26 +106,71 @@ public async Task GivenValidedInput__WhenExecuteAsync_ThenShouldReturnListOfTena
...
                 LastActionCode = "ACC2",
                 LastActionDate = DateTime.Now.AddDays(-2),
-                PrimaryContactPostcode = "test2",
-                PrimaryContactShortAddress = "123DreryLane2",
+                PrimaryContactPostcode = "E5 0AA",
+                PrimaryContactShortAddress = "416 Cherry Street",
                 PropertyRef = "22",
                 Tenure = "LongLease2"
             };
+            var tenancy3 = new TenancyListItem
```
Comment (kjdchapman) [inline LBHTenancyAPITest/Test/UseCases/V2/Search/SearchTenancyUseCaseTests.cs]: This long list of tenancies feels like it's looking for some kind of `factory` object - though I don't know the options for that in `.NET`.

Comment (elena-vi) [reply LBHTenancyAPITest/Test/UseCases/V2/Search/SearchTenancyUseCaseTests.cs]: me neither 🤷‍♀

File: LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs [hunk 6]
```diff
@@ -33,10 +36,34 @@ public async Task<SearchTenancyResponse> ExecuteAsync(SearchTenancyRequest reque
...
+
+            var uniqueTenancyListItems = allTenancyListItems.GroupBy(t => t.TenancyRef)
+                .Where(g => g.Count() == 1)
+                .SelectMany(g => g).ToList();
+
+            var groupsOfDuplicatesTenancyListItems =
+                allTenancyListItems.GroupBy(t => t.TenancyRef)
+                .Where(g => g.Count() > 1);
+
+            foreach (var grouping in groupsOfDuplicatesTenancyListItems)
```
Comment (maysakanoni) [inline LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs]:
```suggestion
            updateUniqueListItems(uniqueTenancyListItems,groupsOfDuplicatesTenancyListItems);
```

Comment (maysakanoni) [reply LBHTenancyAPI/UseCases/V2/Search/SearchTenancyUseCase.cs]: and extract into method like this?
```
private void updateUniqueListItems(List<TenancyListItem> uniqueTenancyListItems,
            IEnumerable<IGrouping<string, TenancyListItem>> groupsOfDuplicatesTenancyListItems)
        {
            foreach (var grouping in groupsOfDuplicatesTenancyListItems)
            {
                var jointTenancies = grouping.ToList();
                var jointTenancy = jointTenancies[0];
                jointTenancies.ForEach(delegate(TenancyListItem dup)
                {
                    if (jointTenancy.PrimaryContactName != dup.PrimaryContactName)
                    {
                        jointTenancy.PrimaryContactName += $" & {dup.PrimaryContactName}";
                    }
                });
                uniqueTenancyListItems.Add(jointTenancy);
            }
        }

**Pull request merged**